### ISSUE TITLE
`Option::is_none_or`: `unnecessary_map_or` lint doesn't work on references

### ISSUE THREAD

**Issue by Azkellas** [rust-lang/rust-clippy#14023]
### Summary

Using `Option::map_or` to emulates `Option::is_none_or` fails to generate a warning when dereferencing.

### Lint Name

unnecessary_map_or

### Reproducer

Running this code

```rust

fn test(a: Option<f32>) {
    println!("{}", a.map_or(true, |f| f < 0.0))
}

fn test_ref(a: &Option<f32>) {
    println!("{}", a.map_or(true, |f| f < 0.0))
}

fn test_ref_mut(a: &mut Option<f32>) {
    println!("{}", a.map_or(true, |f| f < 0.0))
}

fn main() {
    test(Some(1.0));
    test_ref(&Some(1.0));
    test_ref_mut(&mut Some(1.0));
}
```

[link to the playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2024&gist=179d82da68560c282be3a67bd27b223b)

I expected a warning for each case, so three in total

Instead, this happened:

```bash
Checking playground v0.0.1 (/playground)
warning: this `map_or` can be simplified
 --> src/main.rs:3:20
  |
3 |     println!("{}", a.map_or(true, |f| f < 0.0))
  |                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: use is_none_or instead: `a.is_none_or(|f| f < 0.0)`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#unnecessary_map_or
  = note: `#[warn(clippy::unnecessary_map_or)]` on by default

warning: `playground` (bin "playground") generated 1 warning (run `cargo clippy --fix --bin "playground"` to apply 1 suggestion)
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.47s
```

### Version

```text
1.86.0-nightly

(2025-01-17 6067b36314ab5eb2eb47)
```

**Comment by samueltardieu** [comment]
Thanks for the report @Azkellas.
